flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/type/NumericOrDefaultReturnTypeInference.java

```diff
@@ -56,11 +55,7 @@ public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
         List<RelDataType> types = new ArrayList<>();
         for (int i = startTypeIdx; i < nOperands; i++) {
             RelDataType type = opBinding.getOperandType(i);
-            if (SqlTypeUtil.isNumeric(type)) {
-                types.add(type);
-            } else {
-                return opBinding.getOperandType(defaultTypeIdx);
-            }
+            types.add(type);
```

Comment:
Do we really need to skip check for all type?

In the fail case, if(b, 'xx', s), where the second operand is `char(2)`, the third operand is `string`, the it wll mistake the return type is char(2)`, and then in the if code gen, it'll cast all operands to `char(2)`.
Maybe we check 
`if (SqlTypeUtil.isNumeric(type) || SqlTypeUtil.isCharacter(type))` in here?

WDYT?

Reply:
It's just my personal opinion, the reason behind that is the class `NumericOrDefaultReturnTypeInference` is designed for 
- if the operands are Numeric, consider each Numeric and find the leastRestrictive
- Otherwise, it will return the sencond as the return type.

If we skip all checks, the usage of the class `NumericOrDefaultReturnTypeInference` changes.
Then we'll need to modify the class name and java doc.